### ISSUE TITLE
unnecessary variable renames in object destructuring

### ISSUE THREAD

**Issue by heavyk** [terser/terser#446]
**Bug report or Feature request?**

Increases code size. feature request.

**Version (complete output of `terser -V` or specific git commit)**

4.2.1

**`terser` input**

```
console.log(Terser.minify(`
    fn(G, ({h, m}) => {
      return h()
    })
`).code)
// fn(G,({h:f,m:h})=>f())
console.log(Terser.minify(`
    fn(G, ({h, m} = G) => {
      return h()
    })
`).code)
// fn(G,({h:f,m:h}=G)=>f())
```

**`terser` output or error**

```
fn(G,({h:f,m:h})=>f())
fn(G,({h:f,m:h}=G)=>f())
```

**Expected result**

```
// h should not be renamed, because it's not used in the scope
// m should not be renamed -- and, I would argue that it shouldn't even be present
fn(G,({h,m})=>h())
// same as above:  G should be removed because it's not used in the inner scope
fn(G,({h,m})=>h())
```

...so then, I tried the same thing with `pure_getters` and pretty much everything enabled and I got this:

```
// h is still renamed, but m is gone (cause it's not used)
fn(G,({h:f})=>f())
// =G is still not removed (even though it's unused)
// and another problem shows up here (which is a bug): it now thinks 'h' is 'G' in the object
fn(G,({G:f}=G)=>f())
```

**Comment by fabiosantoscode** [comment]
Two reports for the price of one :)

Thank you very much!

I'll have a look at this this week.

**Comment by heavyk** [comment]
there's actually a bug in my code... seems I was thinking js destructuring worked like elixir's pattern matching... what I intended was for `G` to be the argument and then the destructuring. the '=G' bit is bad code (even though it's not a js syntax error (in v8 at least)).

anyway, if I actually do it properly, with a let:
```
console.log(Terser.minify(`
    fn(G, (G) => {
      let {h, m} = G
      return h()
    })
`, opts).code)
```

the code is still does the weird renaming thing though...
```
fn(G,e=>{let{h:n,m:r}=e;return n()});
```
I can get it to remove the unused variables by setting, `pure_getters`..

```
fn(G,e=>{let{h:n}=e;return n()})
```

which, I would have imagined that the minimised argument, 'e' should move the destructure should move up into the arguments, like this:
```
fn(G,({h:n})=>n())
```

---

just wanted to let you know what I put in the original post wasn't even the desired behaviour.. :) cheers!

**Comment by fabiosantoscode** [comment]
Moving the destructure into the arguments is another set of optimizations which would be nice to have, but a bit tricky to do. I'm not sure we want to tackle that before all the outstanding bugs have been fixed and people don't find new ones for a while.

It's a nice optimization and one that I haven't thought about before actually.

**Comment by heavyk** [comment]
yeah sounds good for moving the destructure into the args... someday if I find myself wanting to do something of this nature, I'll have a go at it. 

the main issue is the unnecessary renames. I think that it's *always* trying to rename, even when the unmodified name is shorter than the rename. the rename costs a reassignment so it should only be done if there's a gain to be made.

**Comment by kzc** [comment]
This problem is not as cut and dry as it may appear. It's only suboptimal because of the one letter property names and destructuring variables subsequently used only once. In typical code with longer property names and multiple variable use it's generally not an issue. Mangled names are determined by a character frequency algorithm. Attempting to optimize for this atypical case could potentially negatively impact gzip size for the bundle. I suppose a mangle option could be created to not mangle destructuring variable names of length less than 3, but be disabled by default until which time it can be proven to produce smaller compressed bundles on large real life projects.

Regarding the pure_getters thing - it's necessary because of destructuring of objects that could potentially have getters with side effects. See the test suite for details.

**Comment by fabiosantoscode** [comment]
We might not be able to blanket-optimize everything, but we can optimize the following cases safely:

 - `({foo})=>foo.something` => `a=>a.foo.something`
 - `(something)=>{const {foo}=something;return a}` => `({foo:n})=>n`

And a lot more than this when pure_getters is enabled. But I don't think I want to cater to people using pure_getters since it's not on by default.

**Comment by kzc** [comment]
No need to mess with function parameters in the second case. It would be simpler and shorter if `const {foo}=something;` were transformed to `const foo=something.foo;`. This is a safe transform for all single variable object destructuring without a default value. Then the existing compress machinery would take over to produce:

```
$ echo 'let fn = (something)=>{const {foo}=something;return foo}' | terser -mc passes=1
let fn=e=>{return e.foo};

$ echo 'let fn = (something)=>{const {foo}=something;return foo}' | terser -mc passes=2
let fn=f=>f.foo;
```

**Comment by fabiosantoscode** [comment]
That's a great idea @kzc!

**Comment by kzc** [comment]
> // and another problem shows up here (which is a bug): 
> // it now thinks 'h' is 'G' in the object
> `fn(G,({G:f}=G)=>f())`

@heavyk I can't reproduce that with typical compress/mangle flags:

```
$ terser -V
terser 4.3.1

$ echo 'fn(G,({h,m}=G)=>{return h()});' | terser -mc
fn(G,({h:f,m:h}=G)=>f());

$ echo 'fn(G,({h,m}=G)=>{return h()});' | terser -mc pure_getters
fn(G,({h:f}=G)=>f());

$ echo 'fn(G,({h,m}=G)=>{return h()});' | terser -mc pure_getters,unsafe,passes=3 --toplevel
fn(G,({h:f}=G)=>f());
```

Unless you were using `--mangle-props`:

```
$ cat nc.json && echo
{"vars":{"props":{}},"props":{"props":{"$h":"G","$m":"Z"}}}

$ echo 'fn(G,({h,m}=G)=>{return h()});' | terser -mc --mangle-props --name-cache=nc.json
fn(G,({G:f,Z:h}=G)=>f());

$ echo 'fn(G,({h,m}=G)=>{return h()});' | terser -mc pure_getters --mangle-props --name-cache=nc.json
fn(G,({G:f}=G)=>f());
```

in which case it's working as expected and appears to be correct. Notice in the `nc.json` example above that the property `h` maps to `G` and the property `m` maps to `Z`. The renamed property name `G` does not conflict with the variable name `G` - think of them being in different namespaces.

You really have to know what you're doing when you use `--mangle-props`. You must share the same `--name-cache` for all minified source bundles and be careful not to mangle certain properties beyond the control of your code. There's a reason why it says the following in the documentation: "**Note:** THIS WILL PROBABLY BREAK YOUR CODE."

**Comment by kzc** [comment]
> I tried the same thing with pure_getters and pretty much everything enabled

@fabiosantoscode I don't see any bug here. The problem was likely explained by a misunderstanding of how`--mangle-props` works. 

As far as the code size increase with destructuring variable renaming, it does increase ungzipped minified code size for small contrived examples, but in real life large code bases selectively retaining the variable names which match the destructuring property names always increases the ungzipped code size, and for minified+gzip it has a savings of at most 1/15th of one percent when tested against the ES6 `rollup/dist/rollup.es.js` which makes heavy use of destructuring - i.e., 60 bytes in 97K. It's not worth the extra code complexity in mangling and minification slowdown. The reason why it's a wash is due to the character frequency algorithm already employed in mangle - it's already very effective at producing near-optimal gzippable mangle results - particularly when the shorter renamed destructured variables are subsequently used several times in a given function, as is often the case.
